cache/caffeine-cache/src/main/java/org/jdbi/v3/cache/caffeine/CaffeineCache.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.cache.caffeine;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import org.jdbi.v3.core.cache.JdbiCache;
+import org.jdbi.v3.core.cache.JdbiCacheBuilder;
+import org.jdbi.v3.core.cache.JdbiCacheLoader;
+
+/**
+ * Cache implementation using the caffeine cache library.
+ *
+ * @param <K> The key type.
+ * @param <V> The value type.
+ */
+public final class CaffeineCache<K, V> implements JdbiCache<K, V> {
+
+    private final Cache<K, V> cache;
+    private final boolean isLoadingCache;
+
+    /**
+     * Returns a new {@link JdbiCacheBuilder} which can be used to construct the internal caches.
+     *
+     * @return A {@link JdbiCacheBuilder} instance.
+     */
+    public static JdbiCacheBuilder builder() {
+        return new CaffeineCacheBuilder();
+    }
+
+    CaffeineCache(Caffeine<Object, Object> caffeine, JdbiCacheLoader<K, V> cacheLoader) {
+        this.isLoadingCache = cacheLoader != null;
+
+        this.cache = this.isLoadingCache
+                ? caffeine.build(cacheLoader::create)
+                : caffeine.build();
+    }
+
+    @Override
+    public V get(K key) {
+        if (isLoadingCache) {
+            LoadingCache<K, V> loadingCache = (LoadingCache<K, V>) cache;
```

Comment:
A slight modification would be to replace `isLoadingCache` with a  `LoadingCache<K, V>` field that is set to null if the cache is not loading. Then, you can compare the field against null here, and avoid doing a cast on every get operation. Just a trivial optimization in the hot path, I doubt it matters much.

Reply:
I split the cache classes into two, a cache and a loading cache. Now there is no more branch and checks at all. Applied to caffeine and guava.